File: myLeetCode/Dynamic_Programming/Subsequence_problem/1035. Uncrossed Lines.cpp
// 1035. Uncrossed Lines
// https://leetcode.com/problems/uncrossed-lines/

#include <vector>
// head files
#include <iostream>
using namespace std;

class Solution {
public:
    int maxUncrossedLines(vector<int>& A, vector<int>& B) {
        int m = A.size(), n = B.size();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1));
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (A[i - 1] == B[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        return dp[m][n];
    }
};

// main function
int main() {
    // input
    vector<int> A = {2,5,1,2,5};
    vector<int> B = {10,5,2,1,5,2};
    
    // output
    Solution sol;
    int ans = sol.maxUncrossedLines(A,B);
    cout << ans << endl;
    

}

File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/1035. Uncrossed Lines.py
def maxUncrossedLines(A, B):
    m, n = len(A), len(B)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if A[i - 1] == B[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[m][n]

# 测试用例
print(maxUncrossedLines([1,4,2], [1,2,4]))  # 输出应该是 2
print(maxUncrossedLines([2,5,1,2,5], [10,5,2,1,5,2]))  # 输出应该是 3
